lib/galaxy/workflow/run.py

```diff
@@ -334,10 +334,11 @@ def remaining_steps(
             if not hasattr(step, "module"):
                 self.module_injector.inject(step, step_args=self.param_map.get(step.id, {}))
                 if step_id not in step_states:
-                    template = "Workflow invocation [%s] has no step state for step id [%s]. States ids are %s."
-                    message = template % (self.workflow_invocation.id, step_id, list(step_states.keys()))
-                    raise Exception(message)
+                    raise Exception(
+                        f"Workflow invocation [{self.workflow_invocation.id}] has no step state for step id [{step_id}]. States ids are {list(step_states.keys())}."
```

Comment:
While you're at it I would use log_str:
```suggestion
                        f"Workflow invocation [{self.workflow_invocation.id}] has no step state for step {step.log_str()}. States ids are {list(step_states.keys())}."
```